internal/expr/bindtypes.go

```diff
@@ -44,69 +39,43 @@ func (pe *ParsedExpr) BindTypes(args ...any) (te *TypedExpr, err error) {
 		}
 	}()
 
-	var ti = make(typeNameToInfo)
-
-	// Generate and save reflection info.
-	for _, arg := range args {
-		if arg == nil {
-			return nil, fmt.Errorf("need struct or map, got nil")
-		}
-		t := reflect.TypeOf(arg)
-		switch t.Kind() {
-		case reflect.Struct, reflect.Map:
-			if t.Name() == "" {
-				return nil, fmt.Errorf("cannot use anonymous %s", t.Kind())
-			}
-			info, err := typeinfo.GetTypeInfo(arg)
-			if err != nil {
-				return nil, err
-			}
-			if dupeInfo, ok := ti[t.Name()]; ok {
-				if dupeInfo.Typ() == t {
-					return nil, fmt.Errorf("found multiple instances of type %q", t.Name())
-				}
-				return nil, fmt.Errorf("two types found with name %q: %q and %q", t.Name(), dupeInfo.Typ().String(), t.String())
-			}
-			ti[t.Name()] = info
-		case reflect.Pointer:
-			return nil, fmt.Errorf("need struct or map, got pointer to %s", t.Elem().Kind())
-		default:
-			return nil, fmt.Errorf("need struct or map, got %s", t.Kind())
-		}
+	argInfo, err := typeinfo.GenerateArgInfo(args...)
+	if err != nil {
+		return nil, err
 	}
 
 	var sql bytes.Buffer
 
 	var inCount int
 	var outCount int
 
-	var outputs = make([]typeinfo.Member, 0)
-	var inputs = make([]typeinfo.Member, 0)
+	var outputs = make([]typeinfo.Output, 0)
+	var inputs = make([]typeinfo.Input, 0)
 
-	var typeMemberPresent = make(map[typeinfo.Member]bool)
+	var outputUsed = make(map[typeinfo.Output]bool)
```

Comment:
I actually prefer it as `outputUsed`, its a map rather than a slice so when you look up an output you do `if outputUsed[o]`, or `outputUsed[o] = true` which makes more sense than `if outputsUsed[o]` since you are only checking one.

Reply:
That makes sense as well, I agree with you, let's keep the name as is.